File: app/src/main/java/com/waracle/androidtest/DataSource/DataSources.java
package com.waracle.androidtest.DataSource;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

public interface DataSources<T> {

    String getUrl();

    ConcurrentLinkedQueue<WeakReference<DataListeners<T>>> getDataSourceListeners();

    void run();

    interface DataListeners<R>{

        void onDataRetrieved(R result);

    }
}

File: app/src/main/java/com/waracle/androidtest/dataSource/DataSources.kt
package com.waracle.androidtest.dataSource

import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentLinkedQueue

interface DataSources{
    fun getUrl():String
    fun getDataSourceListeners(): ConcurrentLinkedQueue<WeakReference<DataListeners<Any>>>
    suspend fun run()

    interface DataListeners<R>{
        suspend fun onDataRetrieved(result: R)
    }
}